File: src/main/java/com/tealcube/minecraft/bukkit/mythicdrops/repair/RepairingListener.java
package com.tealcube.minecraft.bukkit.mythicdrops.repair;

/*
 * #%L
 * MythicDrops
 * %%
 * Copyright (C) 2013 - 2015 TealCube
 * %%
 * Permission to use, copy, modify, and/or distribute this software for any purpose with or without fee is hereby
 * granted,
 * provided that the above copyright notice and this permission notice appear in all copies.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
 * INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
 * IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF
 * THIS SOFTWARE.
 * #L%
 */


import com.comphenix.xp.rewards.xp.ExperienceManager;
import com.tealcube.minecraft.bukkit.mythicdrops.api.MythicDrops;
import com.tealcube.minecraft.bukkit.mythicdrops.api.repair.RepairCost;
import com.tealcube.minecraft.bukkit.mythicdrops.api.repair.RepairItem;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.entity.HumanEntity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockDamageEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.*;

public final class RepairingListener implements Listener {

    private MythicDrops mythicDrops;
    private Map<String, ItemStack> repairing;

    public RepairingListener(MythicDrops mythicDrops) {
        this.mythicDrops = mythicDrops;
        repairing = new HashMap<>();
    }

    @EventHandler(priority = EventPriority.MONITOR)
    public void onBlockDamageEvent(BlockDamageEvent event) {
        if (event.isCancelled()) {
            return;
        }
        if (event.getPlayer() == null) {
            return;
        }
        if (!event.getPlayer().hasPermission("mythicdrops.repair")
                || event.getBlock().getType() != Material.ANVIL) {
            return;
        }
        Player player = event.getPlayer();
        if (repairing.containsKey(player.getName())) {
            ItemStack oldInHand = repairing.get(player.getName());
            ItemStack currentInHand = player.getItemInHand();
            if (oldInHand.getType() != currentInHand.getType()) {
                player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                      ".repair-cannot-use"));
                repairing.remove(player.getName());
                return;
            }
            if (oldInHand.getDurability() == 0 || currentInHand.getDurability() == 0) {
                player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                      ".repair-cannot-use"));
                event.setCancelled(true);
                removeMapItem(player);
                return;
            }
            if (!oldInHand.isSimilar(currentInHand)) {
                player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                      ".repair-cannot-use"));
                event.setCancelled(true);
                removeMapItem(player);
                return;
            }
            RepairItem mythicRepairItem = getRepairItem(currentInHand);
            if (mythicRepairItem == null) {
                player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                      ".repair-cannot-use"));
                event.setCancelled(true);
                removeMapItem(player);
                return;
            }
            List<RepairCost> mythicRepairCostList = mythicRepairItem.getRepairCosts();
            if (mythicRepairCostList == null) {
                player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                      ".repair-cannot-use"));
                event.setCancelled(true);
                removeMapItem(player);
                return;
            }
            RepairCost mythicRepairCost =
                    getRepairCost(mythicRepairItem, mythicRepairCostList, player.getInventory());
            if (mythicRepairCost == null) {
                player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                      ".repair-cannot-use"));
                event.setCancelled(true);
                removeMapItem(player);
                return;
            }
            if (!player.getInventory().containsAtLeast(mythicRepairCost.toItemStack(1),
                                                       mythicRepairCost.getAmount())) {
                player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                      ".repair-do-not-have",
                                                                                              new String[][]{
                                                                                                      {"%material%",
                                                                                                       mythicRepairItem
                                                                                                               .toItemStack(
                                                                                                                       1)
                                                                                                               .getType()
                                                                                                               .name()}}
                                                                                             ));
                event.setCancelled(true);
                removeMapItem(player);
                return;
            }
            ExperienceManager experienceManager = new ExperienceManager(player);
            if (!experienceManager.hasExp(mythicRepairCost.getExperienceCost())) {
                player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                      ".repair"
                                                                                                      + ".do-not-have",
                                                                                              new String[][]{
                                                                                                      {"%material%",
                                                                                                       "experience"}}
                                                                                             ));
                event.setCancelled(true);
                removeMapItem(player);
                return;
            }
            experienceManager.changeExp(-mythicRepairCost.getExperienceCost());
            player.setItemInHand(repairItemStack(currentInHand, player.getInventory()));
            removeMapItem(player);
            event.setCancelled(true);
            player.sendMessage(
                    mythicDrops.getConfigSettings().getFormattedLanguageString("command.repair-success"));
            player.updateInventory();
            if (mythicDrops.getRepairingSettings().isPlaySounds()) {
                player.playSound(event.getBlock().getLocation(), Sound.ANVIL_USE, 1.0F, 1.0F);
            }
        } else {
            if (player.getItemInHand().getType() == Material.AIR) {
                return;
            }
            if (player.getItemInHand().getDurability() == 0) {
                return;
            }
            if (getRepairItem(player.getItemInHand()) == null) {
                return;
            }
            if (player.getItemInHand().hasItemMeta()) {
                ItemMeta itemMeta = player.getItemInHand().getItemMeta();
                if (itemMeta.hasLore()) {
                    List<String> lore = new ArrayList<>(itemMeta.getLore());
                    lore.add(ChatColor.BLACK + "Repairing");
                    itemMeta.setLore(lore);
                } else {
                    itemMeta.setLore(Arrays.asList(ChatColor.BLACK + "Repairing"));
                }
                player.getItemInHand().setItemMeta(itemMeta);
            } else {
                ItemMeta itemMeta = Bukkit.getItemFactory().getItemMeta(player.getItemInHand().getType());
                itemMeta.setLore(Arrays.asList(ChatColor.BLACK + "Repairing"));
                player.getInventory().getItemInHand().setItemMeta(itemMeta);
            }
            repairing.put(player.getName(), player.getItemInHand());
            player.sendMessage(mythicDrops.getConfigSettings().getFormattedLanguageString("command" +
                                                                                                  ".repair-instructions"));
        }
    }

    private void removeMapItem(Player player) {
        repairing.remove(player.getName());
        if (player.getItemInHand().hasItemMeta()) {
            ItemMeta itemMeta = player.getItemInHand().getItemMeta();
            if (itemMeta.hasLore()) {
                List<String> lore = removeAllString(itemMeta.getLore(), ChatColor.BLACK + "Repairing");
                itemMeta.setLore(lore);
            }
            player.getItemInHand().setItemMeta(itemMeta);
        }
    }

    private List<String> removeAllString(List<String> list, String toRemove) {
        List<String> newList = new ArrayList<>(list);
        Iterator<String> iterator = newList.iterator();
        while (iterator.hasNext()) {
            String s = iterator.next();
            if (s.equals(toRemove)) {
                iterator.remove();
            }
        }
        return newList;
    }

    private RepairItem getRepairItem(ItemStack itemStack) {
        String displayName = null;
        List<String> lore = new ArrayList<>();
        Material material = itemStack.getType();
        if (itemStack.hasItemMeta()) {
            if (itemStack.getItemMeta().hasDisplayName()) {
                displayName = itemStack.getItemMeta().getDisplayName();
            }
            if (itemStack.getItemMeta().hasLore()) {
                lore = itemStack.getItemMeta().getLore();
            }
        }
        for (RepairItem repItem : MythicRepairItemMap.getInstance().values()) {
            if (repItem.getMaterial() != material) {
                continue;
            }
            if (repItem.getItemName() != null && (displayName == null || !ChatColor
                    .translateAlternateColorCodes('&',
                                                  repItem.getName()).equals(displayName))) {
                continue;
            }
            if (repItem.getItemLore() != null && !repItem.getItemLore().isEmpty()) {
                if (lore == null) {
                    continue;
                }
                List<String> coloredLore = new ArrayList<>();
                for (String s : repItem.getItemLore()) {
                    coloredLore.add(ChatColor.translateAlternateColorCodes('&', s));
                }
                if (!coloredLore.equals(lore)) {
                    continue;
                }
            }
            return repItem;
        }
        return null;
    }

    private RepairCost getRepairCost(RepairItem mythicRepairItem,
                                     List<RepairCost> mythicRepairCostsList, Inventory inventory) {
        RepairCost repCost = null;
        for (RepairCost mythicRepairCost : mythicRepairCostsList) {
            ItemStack itemStack = mythicRepairCost.toItemStack(1);
            if (inventory.containsAtLeast(itemStack, 1)) {
                if (repCost == null) {
                    repCost = mythicRepairCost;
                    continue;
                }
                if (repCost.getPriority() > mythicRepairCost.getPriority()) {
                    repCost = mythicRepairCost;
                }
            }
        }
        return repCost;
    }

    private ItemStack repairItemStack(ItemStack itemStack, Inventory inventory) {
        if (itemStack == null) {
            return null;
        }
        ItemStack repaired = itemStack.clone();
        RepairItem mythicRepairItem = getRepairItem(itemStack);
        if (mythicRepairItem == null) {
            return repaired;
        }
        List<RepairCost> mythicRepairCostList = mythicRepairItem.getRepairCosts();
        if (mythicRepairCostList == null) {
            return repaired;
        }
        RepairCost mythicRepairCost = getRepairCost(mythicRepairItem, mythicRepairCostList, inventory);
        if (mythicRepairCost == null) {
            return repaired;
        }
        if (!inventory.containsAtLeast(mythicRepairCost.toItemStack(1), mythicRepairCost.getAmount())) {
            return repaired;
        }

        inventory.removeItem(mythicRepairCost.toItemStack(mythicRepairCost.getAmount()));

        short currentDurability = repaired.getDurability();
        short newDurability = (short) (currentDurability - repaired.getType().getMaxDurability()
                * mythicRepairCost
                .getRepairPercentagePerCost());
        repaired.setDurability((short) Math.max(newDurability, 0));
        for (HumanEntity humanEntity : inventory.getViewers()) {
            if (humanEntity instanceof Player) {
                ((Player) humanEntity).updateInventory();
            }
        }
        return repaired;
    }

}

File: src/main/kotlin/com/tealcube/minecraft/bukkit/mythicdrops/repair/RepairingListener.kt
/*
 * This file is part of MythicDrops, licensed under the MIT License.
 *
 * Copyright (C) 2020 Richard Harrah
 *
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.tealcube.minecraft.bukkit.mythicdrops.repair

import com.tealcube.minecraft.bukkit.mythicdrops.api.repair.RepairCost
import com.tealcube.minecraft.bukkit.mythicdrops.api.repair.RepairItem
import com.tealcube.minecraft.bukkit.mythicdrops.api.repair.RepairItemManager
import com.tealcube.minecraft.bukkit.mythicdrops.api.settings.SettingsManager
import com.tealcube.minecraft.bukkit.mythicdrops.chatColorize
import com.tealcube.minecraft.bukkit.mythicdrops.containsAtLeast
import com.tealcube.minecraft.bukkit.mythicdrops.getFromItemMetaAsDamageable
import com.tealcube.minecraft.bukkit.mythicdrops.getThenSetItemMetaAsDamageable
import com.tealcube.minecraft.bukkit.mythicdrops.removeItem
import com.tealcube.minecraft.bukkit.mythicdrops.sendMythicMessage
import io.pixeloutlaw.minecraft.spigot.experience.PlayerExperience
import io.pixeloutlaw.minecraft.spigot.mythicdrops.displayName
import io.pixeloutlaw.minecraft.spigot.mythicdrops.lore
import io.pixeloutlaw.minecraft.spigot.orPrerequisite
import io.pixeloutlaw.minecraft.spigot.prerequisites
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.ItemStack
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.UUID

internal class RepairingListener(
    private val repairItemManager: RepairItemManager,
    private val settingsManager: SettingsManager
) : Listener {
    companion object {
        private const val REPAIRING_COOLDOWN_IN_SECONDS: Long = 3
    }

    private val repairingMap = mutableMapOf<UUID, Instant>()

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onPlayerInteractEvent(event: PlayerInteractEvent) {
        // check our prerequisites for repairing
        val prereqs = prerequisites {
            prerequisite { event.useItemInHand() != Event.Result.DENY }
            prerequisite { event.useInteractedBlock() != Event.Result.DENY }
            prerequisite { event.action == Action.LEFT_CLICK_BLOCK }
            prerequisite { settingsManager.configSettings.components.isRepairingEnabled }
            prerequisite { event.clickedBlock?.type == Material.ANVIL }
            prerequisite { event.player.hasPermission("mythicdrops.repair") }
            prerequisite(
                orPrerequisite {
                    // if the player isn't sneaking
                    prerequisite { !event.player.isSneaking }
                    // if the player is sneaking AND we're allowing repairing while sneaking
                    prerequisite {
                        event.player.isSneaking && settingsManager.repairingSettings.isAllowRepairingWhileSneaking
                    }
                }
            )
            prerequisite(
                orPrerequisite {
                    // if the player hasn't repaired at all
                    prerequisite { !repairingMap.containsKey(event.player.uniqueId) }
                    // if the player has repaired recently, is it outside our cooldown?
                    prerequisite {
                        repairingMap.getOrDefault(event.player.uniqueId, Instant.now())
                            .plus(REPAIRING_COOLDOWN_IN_SECONDS, ChronoUnit.SECONDS)
                            .isBefore(Instant.now())
                    }
                }
            )
        }
        if (!prereqs) {
            return
        }

        val player = event.player
        val itemInMainHand = player.equipment?.itemInMainHand
        if (itemInMainHand == null) {
            return
        }

        handleRepairingItemInMainHand(player, itemInMainHand)
    }

    private fun handleRepairingItemInMainHand(player: Player, itemInMainHand: ItemStack) {
        val prereqs =
            prerequisites {
                prerequisite { itemInMainHand.getFromItemMetaAsDamageable { hasDamage() } ?: false }
                prerequisite { !itemInMainHand.type.isAir }
            }
        if (!prereqs) {
            return
        }
        val repairItem = repairItemManager.get().find {
            val matchesMaterial = it.material == itemInMainHand.type
            val matchesName =
                it.itemName?.let { name -> name.chatColorize() == itemInMainHand.displayName } ?: true
            val matchesLore = if (it.itemLore.isNotEmpty()) {
                it.itemLore.chatColorize() == itemInMainHand.lore
            } else {
                true
            }
            matchesMaterial && matchesName && matchesLore
        }
        val repairCost: RepairCost? = repairItem?.let { getRepairCost(it.repairCosts, player) }
        repairAndMessage(player, repairItem, repairCost, itemInMainHand)
    }

    private fun repairAndMessage(
        player: Player,
        repairItem: RepairItem?,
        repairCost: RepairCost?,
        itemInMainHand: ItemStack
    ) {
        val equipment = player.equipment
        if (repairItem == null || equipment == null) {
            player.sendMythicMessage(settingsManager.languageSettings.repairing.cannotUse)
            return
        }
        if (repairCost == null) {
            player.sendMythicMessage(settingsManager.languageSettings.repairing.doNotHave)
            return
        }
        equipment.setItemInMainHand(repairItemStack(itemInMainHand, repairCost))
        player.inventory.removeItem(
            repairCost.material,
            repairCost.itemName,
            repairCost.itemLore,
            repairCost.enchantments,
            repairCost.amount
        )
        player.inventory.viewers.forEach { (it as? Player)?.updateInventory() }
        player.sendMythicMessage(settingsManager.languageSettings.repairing.success)
    }

    private fun getRepairCost(repairCosts: Collection<RepairCost>, player: Player): RepairCost? {
        var repairCost: RepairCost? = null

        for (cost in repairCosts) {
            val inventoryContains = player.inventory.containsAtLeast(
                cost.material,
                cost.itemName,
                cost.itemLore,
                cost.enchantments,
                cost.amount
            )
            val hasExperience = PlayerExperience.hasExp(player, cost.experienceCost)
            if (inventoryContains && hasExperience) {
                if (repairCost == null) {
                    repairCost = cost
                    continue
                }
                // we want the one with the highest priority
                if (cost.priority > repairCost.priority) {
                    repairCost = cost
                }
            }
        }

        return repairCost
    }

    /**
     * Returns a repaired copy of the given ItemStack.
     */
    private fun repairItemStack(itemStack: ItemStack, repairCost: RepairCost): ItemStack {
        val repaired = itemStack.clone()
        val currentDamage = itemStack.getFromItemMetaAsDamageable { damage } ?: 0
        val newDamage =
            currentDamage - (itemStack.type.maxDurability * repairCost.repairPercentagePerCost).toInt().coerceAtLeast(0)
        repaired.getThenSetItemMetaAsDamageable { damage = newDamage }
        return repaired
    }
}
